File: src/main/java/com/inventario/model/Coleta.java
package com.inventario.model;

import java.math.BigDecimal;
import java.sql.Timestamp;
// Formatação de datas centralizada em DateFormatUtils
import com.inventario.util.DateFormatUtils;
import java.util.Objects;

/**
 * Classe que representa uma Coleta no sistema de inventário
 * Registra as coletas realizadas durante o processo de inventário
 */
public class Coleta {
    
    private int id;
    private int idInventario;
    private int idPatrimonio;
    private int idColetor; // Campo mantido para compatibilidade durante migração
    private int idParticipanteInventario; // Nova referência para TABELA_PARTICIPANTE_INVENTARIO
    private Timestamp dataColeta;
    private String statusColeta;
    private String observacaoColeta;
    private String localizacaoAtual;
    private String localizacaoEncontrada;
    private String estadoEncontrado;
    private boolean divergencia;
    private String motivoDivergencia;
    private BigDecimal latitude;
    private BigDecimal longitude;
    private String fotoPatrimonio;
    
    // Campos para itens sem etiqueta
    private boolean semEtiqueta;
    private String descricaoItemSemEtiqueta;
    private String categoriaItemSemEtiqueta;
    
    // Campos transientes para exibição
    private String numeroPatrimonio;
    private String descricaoPatrimonio;
    private String nomeColetor;
    private String descricaoInventario;
    private String nomeSala;  // Nome da sala original do patrimônio (do JOIN)
    
    // Formatação de datas centralizada em DateFormatUtils
    
    // Constantes para status de coleta
    public static final String STATUS_COLETADO = "COLETADO";
    public static final String STATUS_NAO_ENCONTRADO = "NAO_ENCONTRADO";
    public static final String STATUS_DANIFICADO = "DANIFICADO";
    public static final String STATUS_DIVERGENCIA = "DIVERGENCIA";
    
    // Constantes para estados de conservação
    public static final String ESTADO_OTIMO = "Ótimo";
    public static final String ESTADO_BOM = "Bom";
    public static final String ESTADO_REGULAR = "Regular";
    public static final String ESTADO_RUIM = "Ruim";
    public static final String ESTADO_PESSIMO = "Péssimo";
    
    // Construtores
    public Coleta() {
        this.dataColeta = new Timestamp(System.currentTimeMillis());
        this.statusColeta = STATUS_COLETADO;
        this.divergencia = false;
    }
    
    public Coleta(int idInventario, int idPatrimonio, int idColetor) {
        this();
        this.idInventario = idInventario;
        this.idPatrimonio = idPatrimonio;
        this.idColetor = idColetor;
    }
    
    public Coleta(int idInventario, int idPatrimonio, int idParticipanteInventario, boolean useParticipante) {
        this();
        this.idInventario = idInventario;
        this.idPatrimonio = idPatrimonio;
        this.idParticipanteInventario = idParticipanteInventario;
    }
    
    // Getters e Setters
    public int getId() {
        return id;
    }
    
    public void setId(int id) {
        this.id = id;
    }
    
    public int getIdInventario() {
        return idInventario;
    }
    
    public void setIdInventario(int idInventario) {
        this.idInventario = idInventario;
    }
    
    public int getIdPatrimonio() {
        return idPatrimonio;
    }
    
    public void setIdPatrimonio(int idPatrimonio) {
        this.idPatrimonio = idPatrimonio;
    }
    
    public int getIdColetor() {
        return idColetor;
    }
    
    public void setIdColetor(int idColetor) {
        this.idColetor = idColetor;
    }
    
    public int getIdParticipanteInventario() {
        return idParticipanteInventario;
    }
    
    public void setIdParticipanteInventario(int idParticipanteInventario) {
        this.idParticipanteInventario = idParticipanteInventario;
    }
    
    public Timestamp getDataColeta() {
        return dataColeta;
    }
    
    public void setDataColeta(Timestamp dataColeta) {
        this.dataColeta = dataColeta;
    }
    
    public String getStatusColeta() {
        return statusColeta;
    }
    
    public void setStatusColeta(String statusColeta) {
        this.statusColeta = statusColeta;
    }
    
    public String getObservacaoColeta() {
        return observacaoColeta;
    }
    
    public void setObservacaoColeta(String observacaoColeta) {
        this.observacaoColeta = observacaoColeta;
    }
    
    public String getLocalizacaoAtual() {
        return localizacaoAtual;
    }
    
    public void setLocalizacaoAtual(String localizacaoAtual) {
        this.localizacaoAtual = localizacaoAtual;
    }
    
    public String getLocalizacaoEncontrada() {
        return localizacaoEncontrada;
    }
    
    public void setLocalizacaoEncontrada(String localizacaoEncontrada) {
        this.localizacaoEncontrada = localizacaoEncontrada;
    }
    
    public String getEstadoEncontrado() {
        return estadoEncontrado;
    }
    
    public void setEstadoEncontrado(String estadoEncontrado) {
        this.estadoEncontrado = estadoEncontrado;
    }
    
    public boolean isDivergencia() {
        return divergencia;
    }
    
    public void setDivergencia(boolean divergencia) {
        this.divergencia = divergencia;
    }
    
    public String getMotivoDivergencia() {
        return motivoDivergencia;
    }
    
    public void setMotivoDivergencia(String motivoDivergencia) {
        this.motivoDivergencia = motivoDivergencia;
    }
    
    public BigDecimal getLatitude() {
        return latitude;
    }
    
    public void setLatitude(BigDecimal latitude) {
        this.latitude = latitude;
    }
    
    public BigDecimal getLongitude() {
        return longitude;
    }
    
    public void setLongitude(BigDecimal longitude) {
        this.longitude = longitude;
    }
    
    public String getFotoPatrimonio() {
        return fotoPatrimonio;
    }
    
    public void setFotoPatrimonio(String fotoPatrimonio) {
        this.fotoPatrimonio = fotoPatrimonio;
    }
    
    // Getters e Setters para itens sem etiqueta
    public boolean isSemEtiqueta() {
        return semEtiqueta;
    }
    
    public void setSemEtiqueta(boolean semEtiqueta) {
        this.semEtiqueta = semEtiqueta;
    }
    
    public String getDescricaoItemSemEtiqueta() {
        return descricaoItemSemEtiqueta;
    }
    
    public void setDescricaoItemSemEtiqueta(String descricaoItemSemEtiqueta) {
        this.descricaoItemSemEtiqueta = descricaoItemSemEtiqueta;
    }
    
    public String getCategoriaItemSemEtiqueta() {
        return categoriaItemSemEtiqueta;
    }
    
    public void setCategoriaItemSemEtiqueta(String categoriaItemSemEtiqueta) {
        this.categoriaItemSemEtiqueta = categoriaItemSemEtiqueta;
    }
    
    // Campos transientes
    public String getNumeroPatrimonio() {
        return numeroPatrimonio;
    }
    
    public void setNumeroPatrimonio(String numeroPatrimonio) {
        this.numeroPatrimonio = numeroPatrimonio;
    }
    
    public String getDescricaoPatrimonio() {
        return descricaoPatrimonio;
    }
    
    public void setDescricaoPatrimonio(String descricaoPatrimonio) {
        this.descricaoPatrimonio = descricaoPatrimonio;
    }
    
    public String getNomeColetor() {
        return nomeColetor;
    }
    
    public void setNomeColetor(String nomeColetor) {
        this.nomeColetor = nomeColetor;
    }
    
    public String getDescricaoInventario() {
        return descricaoInventario;
    }
    
    public void setDescricaoInventario(String descricaoInventario) {
        this.descricaoInventario = descricaoInventario;
    }
    
    public String getNomeSala() {
        return nomeSala;
    }
    
    public void setNomeSala(String nomeSala) {
        this.nomeSala = nomeSala;
    }
    
    // Métodos utilitários
    public String getDataColetaFormatada() {
        return DateFormatUtils.formatWithDefault(dataColeta, "Não informado");
    }
    
    public String getStatusColetaDescricao() {
        switch (statusColeta != null ? statusColeta : "") {
            case STATUS_COLETADO:
                return "Coletado";
            case STATUS_NAO_ENCONTRADO:
                return "Não Encontrado";
            case STATUS_DANIFICADO:
                return "Danificado";
            case STATUS_DIVERGENCIA:
                return "Divergência";
            default:
                return statusColeta;
        }
    }
    
    public String getDivergenciaDescricao() {
        return divergencia ? "Sim" : "Não";
    }
    
    public boolean temCoordenadas() {
        return latitude != null && longitude != null;
    }
    
    public String getCoordenadasFormatadas() {
        if (temCoordenadas()) {
            return String.format("%.6f, %.6f", latitude, longitude);
        }
        return "";
    }
    
    public boolean temFoto() {
        return fotoPatrimonio != null && !fotoPatrimonio.trim().isEmpty();
    }
    
    public boolean temDivergenciaLocalizacao() {
        return localizacaoAtual != null && localizacaoEncontrada != null &&
               !localizacaoAtual.equals(localizacaoEncontrada);
    }
    
    // Métodos utilitários para itens sem etiqueta
    public boolean isItemSemEtiquetaValido() {
        if (!semEtiqueta) {
            return true; // Se não é sem etiqueta, não precisa validar
        }
        
        // Para itens sem etiqueta, validar campos obrigatórios
        return descricaoItemSemEtiqueta != null && 
               descricaoItemSemEtiqueta.trim().length() >= 10 &&
               categoriaItemSemEtiqueta != null && 
               !categoriaItemSemEtiqueta.trim().isEmpty() &&
               localizacaoEncontrada != null && 
               !localizacaoEncontrada.trim().isEmpty();
    }
    
    public String getDescricaoCompleta() {
        if (semEtiqueta) {
            return descricaoItemSemEtiqueta != null ? descricaoItemSemEtiqueta : "Item sem etiqueta";
        } else {
            return descricaoPatrimonio != null ? descricaoPatrimonio : "";
        }
    }
    
    public String getIdentificacao() {
        if (semEtiqueta) {
            return "SEM ETIQUETA - " + (categoriaItemSemEtiqueta != null ? categoriaItemSemEtiqueta : "Não categorizado");
        } else {
            return numeroPatrimonio != null ? numeroPatrimonio : "";
        }
    }
    
    // Métodos de validação
    public boolean isValida() {
        // Validação básica
        boolean validacaoBasica = idInventario > 0 && idColetor > 0 &&
                                 statusColeta != null && !statusColeta.trim().isEmpty();
        
        if (!validacaoBasica) {
            return false;
        }
        
        // Se é item sem etiqueta, validar campos específicos
        if (semEtiqueta) {
            return isItemSemEtiquetaValido();
        } else {
            // Se não é sem etiqueta, deve ter patrimônio
            return idPatrimonio > 0;
        }
    }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Coleta coleta = (Coleta) obj;
        return id == coleta.id;
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
    
    @Override
    public String toString() {
        return String.format("Coleta{id=%d, patrimonio=%s, coletor=%s, status=%s, data=%s}",
                id, numeroPatrimonio, nomeColetor, statusColeta, getDataColetaFormatada());
    }
}

File: InventarioMobile/app/src/main/java/com/inventario/mobile/data/model/Coleta.kt
package com.inventario.mobile.data.model

/**
 * Modelo simples para Coleta (sem Room)
 * Usado para armazenar dados de coleta localmente
 */
data class Coleta(
    val id: Int? = null,
    val patrimonioId: Int,
    val usuarioId: Int,
    val dataColeta: String,
    val localizacaoAtual: String? = null,
    val localizacaoEncontrada: String? = null,  // Onde o item foi ENCONTRADO durante a coleta
    val estadoEncontrado: String? = "BOM",
    val observacoes: String? = null,
    val fotoPath: String? = null,
    val status: String? = "COLETADO",
    val latitude: Double? = null,
    val longitude: Double? = null,
    val dataCriacao: String? = null,
    val dataAtualizacao: String? = null,
    
    // Campos de divergência
    val divergencia: Boolean = false,
    val motivoDivergencia: String? = null,
    
    // Campos adicionais do servidor
    val numeroPatrimonio: String? = null,
    val descricaoPatrimonio: String? = null,
    val nomeInventario: String? = null,
    val nomeSala: String? = null,  // Localização ORIGINAL do patrimônio
    val nomeColetor: String? = null,
    
    // Campos adicionais para funcionalidade local
    val sincronizado: Boolean = false,
    val tentativasSincronizacao: Int = 0,
    val erroSincronizacao: String? = null,  // v2.6: Motivo do erro de sincronização
    
    // Campos para itens sem etiqueta
    val semEtiqueta: Boolean = false,
    val descricaoItemSemEtiqueta: String? = null,
    val categoriaItemSemEtiqueta: String? = null
) {
    companion object {
        /**
         * Converte DTO para modelo local
         */
        fun fromDto(dto: com.inventario.mobile.data.remote.dto.ColetaDto): Coleta {
            val usuarioId = dto.usuarioIdCamel ?: dto.usuarioId ?: 0
            android.util.Log.d("Coleta.fromDto", "═══════════════════════════════════════")
            android.util.Log.d("Coleta.fromDto", "Convertendo DTO - ID: ${dto.id}")
            android.util.Log.d("Coleta.fromDto", "  numeroPatrimonio: '${dto.numeroPatrimonio}'")
            android.util.Log.d("Coleta.fromDto", "  descricaoPatrimonio: '${dto.descricaoPatrimonio}'")
            android.util.Log.d("Coleta.fromDto", "  localizacaoEncontrada: '${dto.localizacaoEncontrada}'")
            android.util.Log.d("Coleta.fromDto", "  nomeSala: '${dto.nomeSala}'")
            android.util.Log.d("Coleta.fromDto", "  observacaoColeta: '${dto.observacaoColeta}'")
            android.util.Log.d("Coleta.fromDto", "  usuarioIdCamel: ${dto.usuarioIdCamel}, usuarioId: ${dto.usuarioId}")
            android.util.Log.d("Coleta.fromDto", "═══════════════════════════════════════")
            
            return Coleta(
                id = dto.id,
                patrimonioId = dto.patrimonioIdCamel ?: dto.patrimonioId ?: 0,
                usuarioId = usuarioId,
                dataColeta = dto.dataColeta ?: System.currentTimeMillis().toString(),
                localizacaoAtual = dto.nomeSala,  // Localização ORIGINAL do patrimônio
                localizacaoEncontrada = dto.localizacaoEncontrada,  // Onde foi ENCONTRADO
                observacoes = dto.observacaoColeta,
                fotoPath = dto.fotoPath,
                status = dto.statusColeta ?: "coletado",
                latitude = dto.latitude,
                longitude = dto.longitude,
                dataCriacao = dto.dataCriacao ?: System.currentTimeMillis().toString(),
                dataAtualizacao = dto.dataAtualizacao ?: System.currentTimeMillis().toString(),
                nomeInventario = dto.nomeInventario,
                nomeSala = dto.nomeSala,  // Localização ORIGINAL do patrimônio
                nomeColetor = dto.nomeColetor,
                sincronizado = dto.sincronizado ?: true,
                // Garantir que numeroPatrimonio e descricaoPatrimonio sejam sempre populados
                numeroPatrimonio = dto.numeroPatrimonio?.takeIf { it.isNotBlank() } 
                    ?: (dto.patrimonioId?.toString() ?: "0"),
                descricaoPatrimonio = dto.descricaoPatrimonio?.takeIf { it.isNotBlank() } 
                    ?: dto.observacaoColeta?.takeIf { it.isNotBlank() } 
                    ?: "Patrimônio ${dto.patrimonioId ?: 0}",
                // Campos para itens sem etiqueta
                semEtiqueta = dto.semEtiqueta ?: false,
                descricaoItemSemEtiqueta = dto.descricaoItemSemEtiqueta,
                categoriaItemSemEtiqueta = dto.categoriaItemSemEtiqueta
            )
        }
        
        /**
         * Cria uma nova coleta local
         */
        fun createLocal(
            patrimonioId: Int,
            usuarioId: Int,
            observacoes: String? = null,
            latitude: Double? = null,
            longitude: Double? = null,
            numeroPatrimonio: String? = null,
            descricaoPatrimonio: String? = null,
            nomeSala: String? = null
        ): Coleta {
            val currentTime = System.currentTimeMillis().toString()
            return Coleta(
                patrimonioId = patrimonioId,
                usuarioId = usuarioId,
                dataColeta = currentTime,
                observacoes = observacoes,
                latitude = latitude,
                longitude = longitude,
                dataCriacao = currentTime,
                dataAtualizacao = currentTime,
                numeroPatrimonio = numeroPatrimonio,
                descricaoPatrimonio = descricaoPatrimonio,
                nomeSala = nomeSala
            )
        }
    }
    
    /**
     * Converte modelo local para DTO (para enviar ao servidor)
     */
    fun toDto(patrimonio: com.inventario.mobile.data.model.Patrimonio? = null, idInventario: Int? = null): com.inventario.mobile.data.remote.dto.ColetaDto {
        return com.inventario.mobile.data.remote.dto.ColetaDto(
            id = id,
            numeroPatrimonio = patrimonio?.numeroPatrimonio,
            descricaoPatrimonio = patrimonio?.descricao,
            idInventario = idInventario,
            nomeInventario = null, // Será preenchido pelo servidor
            idSala = null, // Será preenchido pelo servidor se necessário
            nomeSala = null, // Será preenchido pelo servidor se necessário
            localizacaoEncontrada = localizacaoAtual,
            estadoEncontrado = "ENCONTRADO", // Estado padrão para coletas manuais
            observacaoColeta = observacoes,
            dataColeta = dataCriacao,
            statusColeta = status,
            nomeColetor = null, // Será preenchido pelo servidor
            semEtiqueta = semEtiqueta, // Usar valor do modelo
            descricaoItemSemEtiqueta = descricaoItemSemEtiqueta,
            categoriaItemSemEtiqueta = categoriaItemSemEtiqueta,
            sincronizado = sincronizado,
            patrimonioId = patrimonioId,
            usuarioId = usuarioId,
            localizacaoAtual = localizacaoAtual,
            observacoes = observacoes,
            fotoPath = fotoPath,
            status = status,
            latitude = latitude,
            longitude = longitude,
            dataCriacao = dataCriacao,
            dataAtualizacao = dataAtualizacao
        )
    }
    
    /**
     * Converte modelo local para MobileColetaRequest (formato correto do servidor)
     */
    fun toMobileColetaRequest(
        patrimonio: com.inventario.mobile.data.model.Patrimonio,
        idInventario: Int,
        deviceId: String? = null,
        appVersion: String? = null
    ): com.inventario.mobile.data.remote.dto.MobileColetaRequest {
        return com.inventario.mobile.data.remote.dto.MobileColetaRequest(
            numeroPatrimonio = patrimonio.numeroPatrimonio,
            idInventario = idInventario,
            usuarioId = usuarioId,
            idSala = patrimonio.salaId?.toInt(),
            localizacaoEncontrada = localizacaoAtual ?: nomeSala,
            estadoEncontrado = this.estadoEncontrado ?: "BOM",
            observacaoColeta = observacoes,
            dataColeta = dataColeta,
            latitude = latitude,
            longitude = longitude,
            fotoPatrimonio = fotoPath,
            semEtiqueta = false,
            descricaoItemSemEtiqueta = null,
            categoriaItemSemEtiqueta = null,
            deviceId = deviceId,
            appVersion = appVersion,
            divergencia = this.divergencia,
            motivoDivergencia = this.motivoDivergencia
        )
    }
}
